Plugins/upenn.cbica.captk.example3/src/internal/QmitkCaPTkExample3View.cpp

```diff
@@ -0,0 +1,205 @@
+#include "QmitkCaPTkExample3View.h"
+
+#include <QMessageBox>
+#include <ExampleInverter.h>
+
+
+const std::string QmitkCaPTkExample3View::VIEW_ID =
+    "upenn.cbica.captk.views.example3";
+
+
+QmitkCaPTkExample3View::QmitkCaPTkExample3View()
+{
+  // ---- General setup operations ----
+}
+
+QmitkCaPTkExample3View::~QmitkCaPTkExample3View()
+{
+
+}
+
+void QmitkCaPTkExample3View::CreateQtPartControl(QWidget *parent)
+{
+  // ---- Setup the basic GUI of this view ----
+  m_Parent = parent;
+  m_Controls.setupUi(parent);
+
+  connect(m_Controls.pushButton_DoStuff, SIGNAL(clicked()), this, SLOT(OnDoStuffButtonClicked()));
+  connect(m_Controls.pushButton_CloneImage, SIGNAL(clicked()), this, SLOT(OnCloneImageButtonClicked()));
+  connect(m_Controls.pushButton_InvertImage, SIGNAL(clicked()), this, SLOT(OnInvertImageButtonClicked()));
+
+}
+
+void QmitkCaPTkExample3View::Activated()
+{
+  // Not yet implemented
+}
+
+void QmitkCaPTkExample3View::Deactivated()
+{
+  // Not yet implemented
+}
+
+void QmitkCaPTkExample3View::Visible()
+{
+  // Not yet implemented
+}
+
+void QmitkCaPTkExample3View::Hidden()
+{
+  // Not yet implemented
+}
+
+/************************************************************************/
+/* protected slots                                                      */
+/************************************************************************/
+
+void QmitkCaPTkExample3View::OnDoStuffButtonClicked()
+{
+    QMessageBox msgError;
+    msgError.setText("Well, now I'm not doing it.");
+    msgError.setIcon(QMessageBox::Critical);
+    msgError.setWindowTitle("No!");
+    msgError.exec();
+}
+
+void QmitkCaPTkExample3View::OnCloneImageButtonClicked()
+{
+	//get datastorage( we use it further down )
+	auto ds = this->GetDataStorage();
+
+	//get selected nodes
+	QList<mitk::DataNode::Pointer> nodes = this->GetDataManagerSelection();
+
+	//we don't handle the case where data is not loaded or more than 1 nodes are selected
+	if (nodes.empty() || nodes.size() > 1)
+	{
+		QMessageBox msgError;
+		msgError.setText("Please load and select a dataset.");
+		msgError.setIcon(QMessageBox::Critical);
+		msgError.setWindowTitle("selection error");
+		msgError.exec();
+	}
+	else
+	{
+		//get first node from list
+		mitk::DataNode *node = nodes.front();
+
+		//clone the selected node
+		mitk::DataNode::Pointer clonedNode = node->Clone();
+
+		//set some new property to cloned node to distinguish it from existing node
+		clonedNode->SetProperty("name", mitk::StringProperty::New("child"));
+
+		//get data from selected node
+		mitk::BaseData* data = node->GetData();
+
+		//set it to cloned node
+		clonedNode->SetData(data);
+
+		//add cloned node to ds
+		ds->Add(clonedNode, node);
+	}
+}
+
+
+void QmitkCaPTkExample3View::OnInvertImageButtonClicked()
+{
+    //get datastorage( we use it further down )
+	auto ds = this->GetDataStorage();
```

Comment:
It is not used further down

Reply:
change this->GetDataStorage() to ds-> towards the end of the function